src/extra/components/adq.py

```diff
@@ -0,0 +1,1302 @@
+
+                if not pulse_period:
+                    raise ValueError('two or more pulses requires to infer '
+                                     'pulse period') from None
+                elif len(pulse_period) > 1:
+                    raise ValueError('more than one period between pulse IDs')\
+                        from None
+
+                pulse_period = int(pulse_period.pop())
+
+        if pulse_period is None:
+            raise ValueError('must pass either pulse_period, repetition_rate, '
+                             'pulse_ids or initialize component with pulse '
+                             'information') from None
+
+        if not fractional:
+            pulse_period = int(round(pulse_period, 1))
+
+        return self._clock_ratio * pulse_period
+
+    def correct_common_mode(self, data, cm_period, baseline, baselevel):
+        """Correct common mode.
+
+        For ADQ digitizers, a common mode is present for every Nth
+        sample due to the parallel readout architecture, typically with
+        N = 8 or N = 16 with interleaving. This method allows to correct
+        this behaviour by computing a baseline for every Nth sample up
+        to the configured period within the baseline region and substract
+        it from every Nth pixel across the entire trace. This will change
+        the native data type to float32 if none is specified explicitly
+        or an output array passed. Note that this type of common mode
+        correction will often not yield useful results for integer types.
+
+        Args:
+            data (ArrayLike): Input data to preprocess, will be
+                converted to np.ndarray currently.
+            out (ArrayLike, optional): Result array, a new is allocated
+                if omitted.
+            cm_period (int, optional): Apply common mode correction with
+                specified sample periodicity if greater than 0, by
+                default 8 or 16 when interleaved.
+            baseline (slice or ArrayLike, optional): Contiguous 1D slice
+                of the trace of each train to determine common mode or
+                baseline or direct baseline data to use, :1000 by default.
+            baselevel (number, optional): ADU value to pull the baseline
+                to, None by default. Note that common mode corrections
+                if enabled always pull the baselevel to zero unless
+                specified otherwise here.
+
+        Returns:
+            out (ndarray): Preprocessed input data, converted to ndarray
+                if a different array was passed.
+        """
+
+        if cm_period < 1:
+            raise ValueError('Common mode must be positive number')
+
+        out = np.zeros_like(data, dtype=np.float32)
+
+        self._correct_cm_by_train(data, out, cm_period, baseline, baselevel)
+
+        return out
+
+    def pull_to_baselevel(self, data, baseline, baselevel):
+        out = np.zeros_like(data, dtype=np.float32)
```

Comment:
Docstring? If you're making this a public method.

Reply:
It looks like my last push was not yet visible, the docstring is there.